04-onion-routing.md

```diff
@@ -729,37 +741,38 @@ errors:
        * [`2`:`len`]
        * [`len`:`channel_update`]
 
-The _final node_ MAY, but _intermediate hops_ MUST NOT, return one of the
+The _final node_ MAY, but an _intermediate hop_ MUST NOT, return one of the
 following errors:
   - if the payment hash has already been paid:
     - MAY treat the payment hash as unknown.
-    - MAY succeed in accepting the HTLC.
+    - MAY succeed in accepting the HTLC [FIXME: is this correct?].
   - if the payment hash is unknown:
     - MUST fail the HTLC:
       1. type: PERM|15 (`unknown_payment_hash`)
   - if the amount paid is less than the amount expected:
     - MUST fail the HTLC.
   - if the amount paid is more than twice the amount expected:
-    - SHOULD fail the HTLC. This allows the origin node to reduce information leakage
-    by altering the amount while not allowing for accidental gross overpayment:
+    - SHOULD [FIXME: MUST?] fail the HTLC. Note: this allows the origin node to reduce
+    information leakage by altering the amount while not allowing for accidental
+    gross overpayment:
       1. type: PERM|16 (`incorrect_payment_amount`)
-  - if the `cltv_expiry` is too low:
+  - if the `cltv_expiry` value is too near the present:
     - MUST fail the HTLC:
       1. type: 17 (`final_expiry_too_soon`)
-  - if the `outgoing_cltv_value` does not match the `cltv_expiry` of the HTLC at
-  the final node:
+  - if the `outgoing_cltv_value` does NOT correspond with the `cltv_expiry` from
+  the final node's HTLC:
     1. type: 18 (`final_incorrect_cltv_expiry`)
     2. data:
        * [`4`:`cltv_expiry`]
-  - if the `amt_to_forward` is greater than the `incoming_htlc_amt` of the HTLC
-  at the final node:
+  - if the `amt_to_forward` is greater than the `incoming_htlc_amt` from the
+  final node's HTLC:
     1. type: 19 (`final_incorrect_htlc_amount`)
     2. data:
        * [`4`:`incoming_htlc_amt`]
 
 ## Receiving Failure Codes
 
-A node:
+A node [FIXME: 'origin', 'return-forwarding', or 'any' node?]:
   - MUST ignore any extra bytes in `failuremsg`.
```

Comment:
Good catch: origin node.  The others don't decode it, so don't know whether there are "extra bytes" or not.

Reply:
fixed